runtime/v2/runner/src/main/java/com/walmartlabs/concord/runtime/v2/runner/vm/WithItemsWrapper.java

```diff
@@ -115,27 +118,97 @@ public void eval(Runtime runtime, State state, ThreadId threadId) {
             throw new IllegalStateException("Can't use non-serializable values in 'withItems': " + item + " (" + item.getClass() + ")");
         }
 
-        Frame loop = Frame.builder()
-                .nonRoot()
-                .build();
+        if (items.isEmpty()) {
+            return;
+        }
+
+        eval(state, threadId, items);
+    }
+
+    protected abstract void eval(State state, ThreadId threadId, ArrayList<Serializable> items);
+
+    static class ParallelWithItems extends WithItemsWrapper {
+
+        protected ParallelWithItems(Command cmd, WithItems withItems, List<String> outVariables) {
+            super(cmd, withItems, outVariables);
+        }
+
+        @Override
+        protected void eval(State state, ThreadId threadId, ArrayList<Serializable> items) {
+            Frame frame = state.peekFrame(threadId);
+
+            // target frame for out variables
+            Frame targetFrame = VMUtils.assertNearestRoot(state, threadId);
+
+            List<Map.Entry<ThreadId, Serializable>> forks = items.stream()
+                    .map(e -> new AbstractMap.SimpleEntry<>(state.nextThreadId(), e))
+                    .collect(Collectors.toList());
+
+            for (int i = 0; i < forks.size(); i++) {
+                Map.Entry<ThreadId, Serializable> f = forks.get(i);
+
+                Frame cmdFrame = Frame.builder()
+                        .nonRoot()
+                        .commands()
+                        .build();
+
+                cmdFrame.setLocal(CURRENT_ITEMS, items);
+                cmdFrame.setLocal(CURRENT_INDEX, i);
+                cmdFrame.setLocal(CURRENT_ITEM, f.getValue());
+
+                // fork will create rootFrame for forked commands
+                Command itemCmd = new ForkCommand(f.getKey(),
+                        new AppendVariablesCommand(outVariables, null, targetFrame),
+                        cmd);
+                cmdFrame.push(itemCmd);
+
+                state.pushFrame(threadId, cmdFrame);
+            }
+
+            state.pushFrame(threadId, Frame.builder()
+                    .commands(new PrepareOutVariables(outVariables, targetFrame))
+                    .nonRoot()
+                    .build());
+
+            frame.push(new JoinCommand(forks.stream().map(Map.Entry::getKey).collect(Collectors.toSet())));
+        }
+    }
+
+    /**
+     * Wraps a command into a loop specified by {@code withItems} option.
+     * Creates a new call frame and keeps the item list, the current item
+     * and the index as frame-local variables.
+     */
+    static class LoopWithItems extends WithItemsWrapper {
```

Comment:
Should call those implementations as `SerialWithItems` and `ParallelWithItems` then?

Reply:
yep